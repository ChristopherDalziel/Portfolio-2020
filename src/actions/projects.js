import database from "../firebase/firebase";

export const createProject = (project) => ({
  type: "CREATE_PROJECT",
  project,
});

export const startCreateProject = (projectData = {}) => {
  return (dispatch, getState) => {
    const { name = "", description = "", technology = "" } = projectData;
    const project = { name, description, technology };
    return database
      .ref("projects")
      .push(project)
      .then((ref) => {
        dispatch(createProject({ id: ref.key, ...project }));
      });
  };
};
