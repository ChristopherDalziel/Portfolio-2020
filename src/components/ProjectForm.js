import React from "react";
import styled from "@emotion/styled";
import { storage } from "../firebase/firebase";
import Button from "../utils/Button";

const ProjectFormContainer = styled.div`
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const ProjectInput = styled.input`
  width: 100%;
  border: none;
  background: transparent;
  margin: 0 0 5px;
  padding-top: 5px;
  padding-bottom: 5px;
  &:focus {
    outline: 0;
    border-bottom: 1px solid #ef626b;
  }
`;

const ProjectTextArea = styled.textarea`
  border: none;
  height: 100px;
  width: 100%;
  max-width: 100%;
  resize: none;
  background: transparent;
  &:focus {
    outline: 0;
    border: 1px solid #ef626b;
  }
`;

class ProjectForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: props.project ? props.project.name : "",
      description: props.project ? props.project.description : "",
      technology: props.project ? props.project.technology : "",
      image: null,
      githubUrl: props.project ? props.project.githubUrl : "",
      url: "",
      progress: 0,
    };
  }

  onNameChange = (e) => {
    const name = e.target.value;
    this.setState(() => {
      return { name };
    });
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => {
      return { description };
    });
  };

  onTechnologyChange = (e) => {
    const technology = e.target.value;
    this.setState(() => {
      return { technology };
    });
  };

  onImageChange = (e) => {
    if (e.target.files[0]) {
      const image = e.target.files[0];
      this.setState(() => ({ image }));
    }
  };

  onGithubUrlChange = (e) => {
    const githubUrl = e.target.value;
    this.setState(() => {
      return { githubUrl };
    });
  };

  handleUpload = (e) => {
    e.preventDefault();
    const { image } = this.state;
    const uploadTask = storage.ref(`images/${image.name}`).put(image);
    uploadTask.on(
      "state_changed",
      (snapshot) => {
        const progress = Math.round(
          (snapshot.bytesTransferred / snapshot.totalBytes) * 100
        );
        this.setState({ progress });
      },
      (error) => {
        console.log(error);
      },
      () => {
        storage
          .ref("images")
          .child(image.name)
          .getDownloadURL()
          .then((url) => {
            this.setState({ url });
          });
      }
    );
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.name) {
      this.setState(() => ({
        error: "Please provide a name ",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        name: this.state.name,
        description: this.state.description,
        technology: this.state.technology,
        githubUrl: this.state.githubUrl,
        url: this.state.url,
      });
    }
  };

  render() {
    return (
      <ProjectFormContainer>
        <form onSubmit={this.onSubmit}>
          <label for="projectName">Project Name:</label>
          <ProjectInput
            autoFocus
            id="projectName"
            name="projectName"
            type="text"
            aria-label="projectName"
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <br />
          <label for="projectDescription">Project Description:</label>
          <ProjectTextArea
            id="projectDescription"
            name="projectDescription"
            type="text"
            aria-label="projectDescription"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          ></ProjectTextArea>
          <br />
          <label for="technologyUsed">Technology Used:</label>
          <ProjectInput
            id="technologyUsed"
            name="technologyUsed"
            type="text"
            aria-label="technologyUsed"
            value={this.state.technology}
            onChange={this.onTechnologyChange}
          />
          <label for="githubUrl">GitHub Url:</label>
          <ProjectInput
            id="githubUrl"
            name="githubUrl"
            type="text"
            aria-label="githubUrl"
            value={this.state.githubUrl}
            onChange={this.onGithubUrlChange}
          />
          <br />
          <input type="file" onChange={this.onImageChange} />
          <br />
          <Button handleClick={this.handleUpload}>Upload</Button>
          <br />
          <progress value={this.state.progress} max="100" />
          <br />
          <button type="submit">Submit Project</button>
        </form>
      </ProjectFormContainer>
    );
  }
}

export default ProjectForm;
