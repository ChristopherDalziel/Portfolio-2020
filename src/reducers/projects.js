const projectReducerDefaultState = [];

const projectReducer = (state = projectReducerDefaultState, action) => {
  switch (action.type) {
    case "CREATE_PROJECT":
      return [...state, action.project];
    case "SET_PROJECTS":
      return action.projects;
    default:
      return state;
  }
};

export default projectReducer;
