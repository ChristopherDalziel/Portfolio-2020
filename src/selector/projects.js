const getVisibleProjects = (projects, { text }) => {
  return projects.filter((project) => {
    const matchText =
      project.name.toLowerCase().includes(text.toLowerCase()) ||
      project.technology.toLowerCase().includes(text.toLowerCase());

    return matchText;
  });
};

export default getVisibleProjects;
