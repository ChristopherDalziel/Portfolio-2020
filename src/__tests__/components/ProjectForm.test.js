import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import projects from "../fixtures/projects";
import ProjectForm from "../../components/ProjectForm";

test("Should render Project Form correctly", () => {
  const wrapper = shallow(<ProjectForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should render Project Form with project data", () => {
  const wrapper = shallow(<ProjectForm project={projects[0]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

// Need to add a display for error to show within component
test("Should render error for a invalid form submission", () => {
  const wrapper = shallow(<ProjectForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
  wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  expect(wrapper.state("error").length).toBeGreaterThan(0);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should set name on input change", () => {
  const value = "test name";
  const wrapper = shallow(<ProjectForm />);
  wrapper.find("#projectName").simulate("change", { target: { value } });
  expect(wrapper.state("name")).toBe(value);
});

test("Should set description on input change", () => {
  const value = "test description";
  const wrapper = shallow(<ProjectForm />);
  wrapper.find("#projectDescription").simulate("change", { target: { value } });
  expect(wrapper.state("description")).toBe(value);
});

test("Should set technology on input change", () => {
  const value = "test technology";
  const wrapper = shallow(<ProjectForm />);
  wrapper.find("#technologyUsed").simulate("change", { target: { value } });
  expect(wrapper.state("technology")).toBe(value);
});

test("Should set githubUrl on input change", () => {
  const value = "test githuburl.com";
  const wrapper = shallow(<ProjectForm />);
  wrapper.find("#githubUrl").simulate("change", { target: { value } });
  expect(wrapper.state("githubUrl")).toBe(value);
});
