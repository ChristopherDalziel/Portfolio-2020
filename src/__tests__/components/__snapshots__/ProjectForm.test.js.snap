// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Project Form correctly 1`] = `
<Styled(div)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="projectName"
    >
      Project Name:
    </label>
    <Styled(input)
      aria-label="projectName"
      autoFocus={true}
      id="projectName"
      name="projectName"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <label
      for="projectDescription"
    >
      Project Description:
    </label>
    <Styled(textarea)
      aria-label="projectDescription"
      id="projectDescription"
      name="projectDescription"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <label
      for="technologyUsed"
    >
      Technology Used:
    </label>
    <Styled(input)
      aria-label="technologyUsed"
      id="technologyUsed"
      name="technologyUsed"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      for="githubUrl"
    >
      GitHub Url:
    </label>
    <Styled(input)
      aria-label="githubUrl"
      id="githubUrl"
      name="githubUrl"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <input
      onChange={[Function]}
      type="file"
    />
    <br />
    <Button
      handleClick={[Function]}
    >
      Upload
    </Button>
    <br />
    <progress
      max="100"
      value={0}
    />
    <br />
    <button
      type="submit"
    >
      Submit Project
    </button>
  </form>
</Styled(div)>
`;

exports[`Should render Project Form with project data 1`] = `
<Styled(div)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="projectName"
    >
      Project Name:
    </label>
    <Styled(input)
      aria-label="projectName"
      autoFocus={true}
      id="projectName"
      name="projectName"
      onChange={[Function]}
      type="text"
      value="Fixture Project 1"
    />
    <br />
    <label
      for="projectDescription"
    >
      Project Description:
    </label>
    <Styled(textarea)
      aria-label="projectDescription"
      id="projectDescription"
      name="projectDescription"
      onChange={[Function]}
      type="text"
      value="Fixture Project 1"
    />
    <br />
    <label
      for="technologyUsed"
    >
      Technology Used:
    </label>
    <Styled(input)
      aria-label="technologyUsed"
      id="technologyUsed"
      name="technologyUsed"
      onChange={[Function]}
      type="text"
      value="Ruby"
    />
    <label
      for="githubUrl"
    >
      GitHub Url:
    </label>
    <Styled(input)
      aria-label="githubUrl"
      id="githubUrl"
      name="githubUrl"
      onChange={[Function]}
      type="text"
      value="http//fakegithuburl.com"
    />
    <br />
    <input
      onChange={[Function]}
      type="file"
    />
    <br />
    <Button
      handleClick={[Function]}
    >
      Upload
    </Button>
    <br />
    <progress
      max="100"
      value={0}
    />
    <br />
    <button
      type="submit"
    >
      Submit Project
    </button>
  </form>
</Styled(div)>
`;

exports[`Should render error for a invalid form submission 1`] = `
<Styled(div)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="projectName"
    >
      Project Name:
    </label>
    <Styled(input)
      aria-label="projectName"
      autoFocus={true}
      id="projectName"
      name="projectName"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <label
      for="projectDescription"
    >
      Project Description:
    </label>
    <Styled(textarea)
      aria-label="projectDescription"
      id="projectDescription"
      name="projectDescription"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <label
      for="technologyUsed"
    >
      Technology Used:
    </label>
    <Styled(input)
      aria-label="technologyUsed"
      id="technologyUsed"
      name="technologyUsed"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      for="githubUrl"
    >
      GitHub Url:
    </label>
    <Styled(input)
      aria-label="githubUrl"
      id="githubUrl"
      name="githubUrl"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <input
      onChange={[Function]}
      type="file"
    />
    <br />
    <Button
      handleClick={[Function]}
    >
      Upload
    </Button>
    <br />
    <progress
      max="100"
      value={0}
    />
    <br />
    <button
      type="submit"
    >
      Submit Project
    </button>
  </form>
</Styled(div)>
`;

exports[`Should render error for a invalid form submission 2`] = `
<Styled(div)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="projectName"
    >
      Project Name:
    </label>
    <Styled(input)
      aria-label="projectName"
      autoFocus={true}
      id="projectName"
      name="projectName"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <label
      for="projectDescription"
    >
      Project Description:
    </label>
    <Styled(textarea)
      aria-label="projectDescription"
      id="projectDescription"
      name="projectDescription"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <label
      for="technologyUsed"
    >
      Technology Used:
    </label>
    <Styled(input)
      aria-label="technologyUsed"
      id="technologyUsed"
      name="technologyUsed"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      for="githubUrl"
    >
      GitHub Url:
    </label>
    <Styled(input)
      aria-label="githubUrl"
      id="githubUrl"
      name="githubUrl"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <input
      onChange={[Function]}
      type="file"
    />
    <br />
    <Button
      handleClick={[Function]}
    >
      Upload
    </Button>
    <br />
    <progress
      max="100"
      value={0}
    />
    <br />
    <button
      type="submit"
    >
      Submit Project
    </button>
  </form>
</Styled(div)>
`;
